README.MD
<div align="center">

  <img src="./res/icons/bilibili.svg" height="100px" width="100px"/>

  <h3>哔哩哔哩 多平台版</h3>
  <br>

----

  这是哔哩哔哩 多平台版
  
  SVG图标来自 [@Peternal](https://github.com/Peternal)

  <br>
</div>

## 介绍

本项目用到了：

1. 反混淆
2. 调试
3. 脑子

本项目完全开源，且没有任何代码加密操作，如有疑虑请自行审查代码或停止使用相关文件。

## 特点

1. 没有讨厌的标题栏
2. 支持小分辨率屏幕全屏
3. 检查更新支持
4. 有能够关闭程序的菜单
5. 支持漫游
6. 外区搜索
7. [弹幕共享](docs/help/弹幕共享.MD)

## 使用方法

### 自动化构建

请到 Release 页面下载：

<https://github.com/Howard20181/bilibili-multiplatform/releases>

即时构建版本：

<https://github.com/Howard20181/bilibili-multiplatform/releases/tag/continuous>

### 手动构建

1. 拉取代码

   ```shell
   git clone https://github.com/Howard20181/bilibili-multiplatform.git
   cd bilibili-multiplatform
   ```

2. 安装

   ```shell
   tools/setup-bilibili
   ```

3. 启动

   ```shell
   bin/bilibili
   ```

### AppImage 安装

请使用 [AppImageLauncher](https://github.com/TheAssassin/AppImageLauncher)

### 额外补充

如果阁下不喜欢电脑多装一个 Electron，可以自行提取发布版的 `app.asar` 并使用已安装 Electron 启动，建议的 Electron 版本是：`21.3.3`

## Flag参数配置

支持的参数列表：
<https://www.electronjs.org/docs/latest/api/command-line-switches>

1. 创建 `flags` 文件

   在 `~/.config/bilibili` 目录下创建 `bilibili-flags.conf` 文件

2. 填写配置

   ```shell
   --disable-gpu
   --key=value
   ```

## 开发者工具

现已支持自行打开开发者工具，方法如下：

尝试按下<kbd>F12</kbd>，如果没有反应就是没开；

登录界面有两层，内层右键打开，外层<kbd>F12</kbd>打开。

## 预览

![推荐](res/screenshots/1.png)

![直播](res/screenshots/live.png)

![检查更新](res/screenshots/update.png)

![退出](res/screenshots/exit.png)

![漫游](res/screenshots/roaming.png)

## 感谢以下项目

1. [BiliRoaming](https://github.com/yujincheng08/BiliRoaming)
2. [解除B站区域限制](https://github.com/ipcjs/bilibili-helper)
3. [BiliBili Linux版](https://github.com/msojocs/bilibili-linux)

## 免责声明

哔哩哔哩客户端版权归上海宽娱数码科技有限公司所有；

对于使用本项目产生的额外问题，如账户封禁被盗等，维护者不对此负责，请谨慎使用；


CHANGELOG.MD
# 🌈 v1.13.0-2 / 2023-12-26

## 🐞 修复

- 外区番剧: 修复港澳台地区和泰区番剧播放
- 外区番剧: 泰区番剧无法切换到 4K
- 外区番剧: 修复 Access Token 失效的问题
- 内核: 更新 Electron 到 27.1.3
- 其他: 同步上游，更新客户端版本到 1.13.0

# 🌈 v1.13.0-1 / 2023-12-24 上游更新

## 🚀 功能

- bilibili: update to v1.13.0.3194

# 🌈 v1.12.5-5 / 2023-12-17

## 🐞 修复

- 外区番剧: 修复部分泰区番剧播放
- 外区番剧: 修复泰区番剧无法切换为1080P高码率的问题
- 控制台: 未登录或登录信息过期时打印日志
- 控制台: 修复右键菜单无法弹出的问题

# 🌈 v1.12.5-4 / 2023-11-30

## 🐞 修复

- 外区番剧: 修复泰区番剧播放
- 外区番剧: CORS 绕过基于域名而不是页面，不再影响其他请求
- 内核: 更新 Electron 到 27.1.2
- 控制台: 减少打印的日志
- 持续集成: 更新构建工具版本

# 🌈 v1.12.5-3 / 2023-11-27

## 🐞 修复

- 外区番剧: 视频质量每次被重置为1080P的问题
- 外区番剧: 视频编码锁定为 AVC 的问题
- 外区番剧: 播放器侧栏追番按钮状态不一致的问题
- 外区番剧: 番剧每话标题 undefined 的问题
- 播放器增强: 播放器增强偶尔不加载的问题

# 🌈 v1.12.5-2 / 2023-11-05

## 🐞 修复

- AccessToken：在某些情况下HD无法登录

# 🌈 v1.12.5-1 / 2023-11-04

## 🐞 修复

- 启动：无法启动

## 🚀 功能

- bilibili: update to v1.12.5.3101

# 🌈 v1.12.3-1 / 2023-10-30

## 🐞 修复

- 启动：处理总有个不影响使用的报错 `require(...) is not a function`
- 控制台: 日志被屏蔽的问题

## 🚀 功能

- bilibili: update to v1.12.3.3068
- 开发者控制台: 使用<kbd>F12</kbd>，而不是配置文件
- hook: 对 `Module` 进行hook，替换 `require`
- 外区番剧：新增HD扫码获取AccessToken，旧的获取方法已被官方禁用 (感谢 @cxw620 提供的帮助) #91

# 🌈 v1.12.1-3 / 2023-10-16

## 🐞 修复

- 用户主页: 用户背景与投稿不显示 #90

# 🌈 v1.12.1-2 / 2023-10-01

## 🐞 修复

- 番剧出差: 出差的账户信息加载失败 #86
- 动态：动态中的港澳台番剧无法播放 #86
- 番剧：字幕繁体转简体 #87

# 🌈 v1.12.1-1 / 2023-09-30

## 🚀 功能

- bilibili: update to v1.12.1.2862

# 🌈 v1.12.0-2 / 2023-09-23

## 🐞 修复

- 番剧搜索: 外区番剧无法搜索
- 直播: 直播页面加载不全

# 🌈 v1.12.0-1 / 2023-09-16

## 🚀 功能

- bilibili: update to v1.12.0.2837
- devtools: 支持通过配置打开开发者工具

# 🌈 v1.11.4-4 / 2023-09-10

## 🐞 修复

- settings: 漫游设置页面有概率无法加载

## 🚧 其它

- extension: 优化扩展加载

# 🌈 v1.11.4-3 / 2023-09-03

## 🐞 修复

- search: 无法搜索到外区番剧

## 🚧 其它

- extension: 重构拓展中的脚本加载逻辑

# 🌈 v1.11.4-2 / 2023-09-03

## 🐞 修复

- 用户动态无法显示(该问题是由于多个实例窗口共用一个`HookXMLHttpRequest`)

## 🚧 其它

- HookXMLHttpRequest: 重构，以方便维护

# v1.11.4-1 / 2023-08-27

- fix: 跨区搜索异常
- update: bilibili to v1.11.4.2706

# v1.10.1-4 / 2023-06-03

- fix: 修复深色模式无法切换问题 #72
- perf: 优化漫游按钮样式 #72

# v1.10.1-3 / 2023-06-03

- feat: flag参数识别加载

# v1.10.1-2 / 2023-06-03

- fix: 弹幕ext有概率无法加载
- fix: 应用启动有概率失败

# v1.10.1-1 / 2023-05-10

- update: electron to v21.3.3
- update: bilibili to v1.10.1.2297

# v1.9.2-1 / 2023-03-25

- update: bilibili to v1.9.2.2092
- feat: 视频硬解支持
- fix: win无法搜索其它区域番剧 #64
- fix: 东南亚视频播放

# v1.9.1-3 / 2023-02-11

- fix: 东南亚提示未登录，导致不能切换清晰度
- fix: 弹幕池相关操作，时间调整，弹幕替换
- fix: Windows不能播放部分视频

# v1.9.1-2 / 2023-02-09

- fix: 番剧出差信息页面
- fix: 港澳台番剧搜索
- fix: 东南亚播放异常

# v1.9.1-1 / 2023-02-07

- update: bilibili to v1.9.1.2031
- fix: 搜索没有任何番剧

# v1.9.0-2 / 2023-01-19

- fix: can not check for update

# v1.9.0-1 / 2023-01-18

- update: bilibili to v1.9.0.1982

# v1.8.2-1 / 2022-12-17

- update: bilibili to v1.8.2.1852

# v1.7.4-1 / 2022-11-24

- update: bilibili to v1.7.4.1613

# v1.7.2-1 / 2022-11-05

- update: bilibili to v1.7.2.1559

# v1.6.1-3 / 2022-09-24

- fix: 港澳台搜索（取决于服务器是否支持cookie转发）

# v1.6.1-2 / 2022-09-24

- fix: 标题栏重现
- fix: 漫游页面无法加载
- fix: 搜索HOOK

# v1.6.1-1 / 2022-09-22

- bilibili: update to v1.6.1.1379

# v1.5.0-3 / 2022-09-01

- fix: 搜索没有任何结果，港澳台仍然有问题。

# v1.5.0-2 / 2022-08-28

- fix: 漫游页面无法加载

# v1.5.0-1 / 2022-08-21

- bilibili: update to v1.5.0.1184
- fix: 漫游按钮异常

# v1.4.1-1 / 2022-08-07

- bilibili: update to v1.4.1.1055
- fix: extension of player

# v1.3.1-2 / 2022-07-24

- fix: 东南亚部分番剧集可能不能播放

# v1.3.1-1 / 2022-07-15

- bilibili: update to v1.3.1.922

# v1.2.3-2 / 2022-07-10

- fix: 东南亚搜索可能无结果
- fix: 东南亚部分番剧集可能不能播放

# v1.2.3-1 / 2022-06-28

- bilibili: update to v1.2.3.809

# v1.2.2-1 / 2022-06-24

- bilibili: update to v1.2.2.796

# v1.2.1-1 / 2022-06-17

- bilibili: update to v1.2.1.724

# v1.2.0-1 / 2022-06-17

- perf: 对搜索进行超时限制
- fix: 修正东南亚图片显示异常
- fix: 修正东南亚视频可能播放失败的问题
- fix: 修正弹幕时间轴调整无效的问题
- bilibili: update to v1.2.0.700

# v1.1.3-4 / 2022-06-06

- feat: 弹幕支持添加港澳台
- update: 子版本更新 494 -> 542
- fix: 播放东南亚视频后其它地区视频不能播放

# v1.1.3-3 / 2022-06-03

- feat: 增加1.75倍速
- fix: 中文域名报错的问题
- feat: 弹幕菜单图标
- feat: 获取弹弹play弹幕可选是否获取关联的第三方

# v1.1.3-2 / 2022-05-30

- perf: 优化拓展注入
- fix: 修复任务栏图标问题
- feat: 弹幕时间轴调整

# v1.1.3-1 / 2022-05-27

- feat: 漫游页面兼容黑暗模式
- fix: 东南亚播放链接获取失败
- update: bilibili v1.1.3

# v1.1.2-3 / 2022-05-24

- fix: 直播间不能关闭
- fix: 界面太小，需要二次拉伸的问题
- feat: 外部弹幕填充

# v1.1.2-2 / 2022-05-21

- fix: vue element-plus 可能加载失败的处理

# v1.1.2-1 / 2022-05-20

- update: bilibili v1.1.2


extensions/area_unlimit/hook/player.js
const pacLink = localStorage.pacLink || ""
let result = ""
if(pacLink.length > 0)
  result = biliBridgePc.callNativeSync('config/roamingPAC', pacLink);
if(result === 'error')localStorage.pacLink = ""
log.info("[hook]: player.js");

const sleep = (ms) => {
  return new Promise((resolve, reject) => {
    setTimeout(resolve, ms)
  })
}
(()=>{

  const HTTP = {
    get(url) {
      return new Promise((resolve, reject) => {
        const Http = new XMLHttpRequest()
        Http.open('GET', url)
        Http.send()
        Http.onloadend = e => {
          resolve(Http)
        }
        Http.onerror = e => reject
      })
    }
  }
  const BilibiliAPI = {
    getEpDetails: (seasonId, epId)=>{
      const api = new BiliBiliApi()
      return api.getSeasonSectionBySsId(seasonId || "")
      .then(seasonInfo=>{
        log.log('seasonInfo:', seasonInfo)
        if(seasonInfo.code !==0)return Promise.reject(seasonInfo)

        const ep = seasonInfo.result.main_section.episodes.filter(ep=>ep.ep_id===parseInt(epId))
        if(ep.length === 0)return Promise.reject("剧集查找失败, target:", epId)
        return Promise.resolve(ep[0])
      })
    }
  }
  const DandanAPI = {
    getComment(epId, withRelated=false){
      const url = `https://api.dandanplay.net/api/v2/comment/${epId}?withRelated=${withRelated}`
      return HTTP.get(url).then(res=>{
        const resp = JSON.parse(res.responseText || "{}")
        return Promise.resolve(resp.comments || [])
      })
    }
  }
  const SearchAPI = {
    bilibili: (str)=>{
      const url = `https://api.bilibili.com/x/web-interface/search/type?__refresh__=true&_extra=&context=&page=1&page_size=12&order=&duration=&from_source=&from_spmid=333.337&platform=pc&device=win&highlight=1&single_column=0&keyword=${str}&search_type=media_bangumi`
      return HTTP.get(url).then(res=>{
        const resp = JSON.parse(res.responseText)
        log.log('bilibili:', resp)
        const bangumiList = []
        const result = resp.data?.result ?? []
        log.log('result:', result)
        for(let bangumi of result){
          let children = []
          if(!bangumi.eps)continue;
          for(let ep of bangumi.eps){
            let title = ep.title || ep.org_title
            title = title.replace(/<.*?>/g, '')
            children.push({
              label: title,
              value: ep.id
            })
          }
          bangumiList.push({
            label: bangumi.title.replace(/<.*?>/g, ''),
            value: bangumi.pgc_season_id,
            children
          })
        }
        return Promise.resolve(bangumiList)
      })
    },
    dandanplay: (str)=>{
      const url = `https://api.dandanplay.net/api/v2/search/episodes?anime=${str}`
      return HTTP.get(url).then(res=>{
        const resp = JSON.parse(res.responseText)
        log.log('dandanplay:', resp)
        const bangumiList = []
        const result = resp?.animes ?? []
        log.log('dandanplay result:', result)
        for(let anime of result){
          let children = []
          for(let ep of anime.episodes){
            children.push({
              label: ep.episodeTitle,
              value: ep.episodeId
            })
          }
          bangumiList.push({
            label: anime.animeTitle,
            value: anime.animeId,
            children
          })
        }
        return Promise.resolve(bangumiList)
      })
    }
  }
  const HandleResult = {
    bilibili: async (options)=>{
      log.log('bilibili options:', options)
      const epDetails = await BilibiliAPI.getEpDetails(...options)
      log.log('getEpDetails:', epDetails)

      // 弹幕池操作
      danmakuManage.rootStore.configStore.reload.cid = epDetails.cid
      // danmakuManage.rootStore.configStore.reload.aid = epDetails.aid
      danmakuManage.danmaku.manager.dataBase.timeLine.list = []
      // 清空当前屏幕的弹幕
      danmakuManage.danmaku.clear()
      // 重载弹幕
      danmakuManage.danmakuStore.loadDmPbAll(true)
      return Promise.resolve("操作成功")
    },
    dandanplay: async (options, data)=>{
      log.log('dandanplay options:', options)
      const comments = await DandanAPI.getComment(options[1], data.dandanplayWithRelated || true)
      log.log('getComment:', comments)
      const result = []
      const nowTime = new Date().getTime()/1000
      for(let comment of comments){
        const p = comment.p.split(',')
        // 出现时间,模式,颜色,用户ID
        const time = parseFloat(p[0])
        const mode = parseInt(p[1])
        const color = parseInt(p[2])
        result.push({
          attr: -1,
          color,
          date: nowTime,
          mode,
          pool: 0,
          renderAs: 1,
          size: 25,
          text: comment.m,
          stime: time,
          weight: 1,
        })
      }
      /**
       * attr: -1
        color: 16777215
        date: 1653221671
        dmid: "1058059079576006912"
        effect: {}
        mode: 1
        pool: 0
        renderAs: 1
        size: 25
        stime: 8.295
        text: "好！"
        uhash: "c515e33f"
        weight: 1
       */
      // 弹幕池操作
      danmakuManage.danmaku.reset()
      const list = danmakuManage.danmaku.manager.dataBase.timeLine.list
      if(data.actionMode === "1") {
          list.splice(0, list.length)
      }
      list.push(...result)
      // 清空当前屏幕的弹幕
      danmakuManage.danmaku.clear()
      // danmakuManage.danmakuStore.loadDmPbAll(true)
      
      return Promise.resolve(`成功加载${comments.length}条弹幕`)
    }
  }
  const UI = (()=>{
    const init = ()=>{
      log.log("init")
      const appContainer = document.querySelector("#app > div > div.app_player--header.flex_between.draggable.db_click_max")
      const page = document.createElement('div')
      page.className = "msojocs-player-settings"
      page.id = "msojocs-player-settings"
      page.style.display = "none"
      appContainer.lastChild.prepend(page)
      const loadStatus = document.createElement('span')
      loadStatus.style.color = "red"
      loadStatus.style.fontSize = "xxx-large"
      page.appendChild(loadStatus)

      function createVueJS() {
        let ele = document.createElement('script');
        ele.src = "https://lib.baomitu.com/vue/3.2.31/vue.global.prod.min.js";
        return ele
      }
    
      function createElementPlusJS() {
        let ele = document.createElement('script');
        ele.src = "https://lib.baomitu.com/element-plus/2.2.0/index.full.min.js";
        return ele
      }
      let vue = createVueJS()
      loadStatus.textContent = "[1/2]加载vue"
      vue.onerror = (e) => {
        const reload = document.createElement('button')
        reload.textContent = "重载vue"
        reload.className = "vui_button about-button mr_sm"
        reload.onclick = () => {
          vue.remove()
          let vueNew = createVueJS()
          vueNew.onload = vue.onload
          vueNew.onerror = vue.onerror
          page.prepend(vueNew)
          reload.remove()
        }
        loadStatus.append(reload)
      }
      page.prepend(vue)
    
      let ele = createElementPlusJS()
      ele.onerror = () => {
        const reload = document.createElement('button')
        reload.textContent = "重载ele"
        reload.className = "vui_button about-button mr_sm"
        reload.onclick = function () {
          ele.remove()
          let eleNew = createElementPlusJS()
          eleNew.onload = ele.onload
          eleNew.onerror = ele.onerror
          reload.remove()
          page.prepend(eleNew)
        }
    
        loadStatus.children.length === 0 && loadStatus.append(reload)
      }
      vue.onload = (e) => {
        loadStatus.textContent = "[2/2]加载element-plus"
        loadStatus.children.length === 1 && loadStatus.children[0].remove()
        ele.onerror()
        page.prepend(ele)
      }
    
      document.addEventListener('ROAMING_sendURL', async function (e) {
        // e.detail contains the transferred data (can be anything, ranging
        // from JavaScript objects to strings).
        // Do something, for example:
        log.log('player ROAMING_sendURL:', e.detail);
        if(e.detail.includes("PlayerEnhance")){
          const roamingHTML = await HTTP.get(e.detail)
          const container = document.createElement('div')
      
          container.innerHTML = roamingHTML.responseText
          ele.onload = () => {
            loadStatus.textContent = ""
            page.className = ""
            loadPage()
          }
          page.appendChild(container)
        }
      });
      document.dispatchEvent(new CustomEvent('ROAMING_getURL', {
        detail: 'PlayerEnhance'
      }));
    }
    let changeShow = ()=>{
      const page = document.getElementById("msojocs-player-settings")
      if(page){
        page.style.display = page.style.display === "none" ? "" : "none"
      }
    }
    const loadPage = ()=>{
      log.log("Vue Start")
      const App = {
        data() {
          return {
            activeName: "bilibili",
            searchStr: "",
            searchResult: [],
            selectOptions: null,
            settingsVisible: false,
            damakuMode: "1",
            dmTimelineDrawer: false,
            moveFactor: 0,
            dandanplayWithRelated: true,
          };
        },
        created() {
          log.log('vue created')
          document.getElementById("player-ext-settings").onclick = ()=>{
            this.settingsVisible = !this.settingsVisible
          }
          this.settingsVisible = document.getElementById("msojocs-player-settings").style.display !== "none"
          document.getElementById("msojocs-player-settings").style.display = ""
          UI.dmTimeline = ()=>{
            this.dmTimelineDrawer = !this.dmTimelineDrawer
          }
        },
        methods: {
          doSearch: function(){
            SearchAPI[this.activeName](this.searchStr)
            .then(resp=>{
              this.searchResult = resp || []
            })
          },
          doConfirm: function(){
            log.log('selectOptions', this.selectOptions)
            let data = {}
            switch (this.activeName) {
              case "dandanplay":
                data.damakuMode = this.damakuMode
                data.dandanplayWithRelated = this.dandanplayWithRelated
                break;
            
              default:
                break;
            }
            HandleResult[this.activeName](this.selectOptions, data)
            .then(res=>{
              this.$message({
                message: res,
                type: 'success'
              })
            }).catch(err=>{
              log.error('出现错误:', err)
              this.$message({
                message: "出现错误",
                type: 'error'
              })
            })
            this.settingsVisible = !this.settingsVisible
          },
          /**
           * 弹幕时间轴调整
           * 正值：弹幕提前
           * 负值：弹幕延后
           * 
           * @param {Number} time 
           */
          dmTimelineMove: function(time){
            this.moveFactor += time
            log.log('dmTimelineMove:', time, this.moveFactor)
            const list = danmakuManage.danmaku.manager.dataBase.timeLine.list
            list.forEach(dm => {
              dm.stime += time
            });

            list.forEach(dm => {
              dm.stime += time
            });
          }
        }
      };
      const app = Vue.createApp(App);
      app.use(ElementPlus);
      app.mount("#player-settings-ext");
    }
    return {
      init,
      changeShow,
    }
  })()

  window.onload = async ()=>{
    log.log("====onload====")
    let headerLeft = null

    for (let i = 0; true; i++) {
      headerLeft = document.querySelector("#app > div > div.app_player--header.flex_between.draggable.db_click_max > div.app_player--header-left")
      if (headerLeft !== null) break
      log.error('头部元素未找到！', i)
      await sleep(1000)
      if (i > 20) {
        return
      }
    }
    // 创建菜单元素
    const playerExtPage = document.createElement('span')
    playerExtPage.innerHTML = `<svg width="26" height="26" style="margin-right:5px" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="1305" ><path d="M1023.8 604.7c0 160.9-129 291.3-287.9 291.3H256.1C113.3 888.5 0.2 781.3 0.2 650.7c0-99.4 65.6-185 159.9-223.5 19.7-8 33.8-25.5 38.2-46.3C224.9 254.9 340.8 160 480 160c102.6 0 192.4 51.5 243.4 129 10 15.3 26.1 25.5 44.2 28.2 145.2 22 256.2 142.1 256.2 287.5z" p-id="1306" fill="#ffffff"></path></svg>弹幕Ext`
    playerExtPage.id = "player-ext-settings"
    playerExtPage.className = "app_player--header-home no_drag"
    playerExtPage.onclick = ()=>{
      UI.changeShow()
    }
    headerLeft.appendChild(playerExtPage)
    const dmTimeline = document.createElement('span')
    dmTimeline.innerHTML = `<svg  width="26" height="26" style="margin-right:5px"fill="none" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M15.25 13c.966 0 1.75.783 1.75 1.75v4.503a1.75 1.75 0 0 1-1.75 1.75H3.75A1.75 1.75 0 0 1 2 19.253V14.75c0-.967.783-1.75 1.75-1.75h11.5ZM21 14.895v5.355a.75.75 0 0 1-1.494.102l-.006-.102v-5.344a3.003 3.003 0 0 0 1.5-.01Zm-.75-4.803a1.908 1.908 0 1 1 0 3.816 1.908 1.908 0 0 1 0-3.816Zm-5.005-7.095c.967 0 1.75.783 1.75 1.75V9.25a1.75 1.75 0 0 1-1.75 1.75h-11.5a1.75 1.75 0 0 1-1.75-1.75V4.747a1.75 1.75 0 0 1 1.607-1.745l.143-.005h11.5ZM20.25 3a.75.75 0 0 1 .743.648L21 3.75v5.345a3.004 3.004 0 0 0-1.5-.01V3.75a.75.75 0 0 1 .75-.75Z" fill="#ffffff" class="fill-212121"></path></svg>弹幕时间轴`
    dmTimeline.style.marginLeft = "5px"
    // dmTimeline.id = "player-timeline-settings"
    dmTimeline.className = "app_player--header-home no_drag"
    dmTimeline.onclick = ()=>{
      UI.dmTimeline && UI.dmTimeline()
    }
    headerLeft.appendChild(dmTimeline)
    UI.init()
    // 添加按钮到页面

    getScript('https://greasyfork.org/scripts/430412-chinese-conversion-api/code/Chinese%20Conversion%20API.js')

  }
  // 1.75倍速
  let rate175check = setInterval(()=>{
    // log.log('1.75倍速')
    try{
      const speedRate = window.danmakuManage.nodes.controlBottomRight.querySelector('.cpx-player-ctrl-playbackrate-menu > li:nth-child(1)')
      const rate175 = document.createElement('li')
      rate175.className = "cpx-player-ctrl-playbackrate-menu-item"
      rate175.dataset.value = "1.75"
      rate175.textContent = "1.75x"
      speedRate.after(rate175)
      clearInterval(rate175check)
    }catch(err){
      // log.error('添加1.75倍速失败：', err)
    }
  }, 1000)
})()